loris/utils.py

```diff
@@ -41,3 +42,31 @@ def symlink(src, dst, force=True):
         os.unlink(dst)
 
     os.symlink(src, dst)
+
+
+def safe_rename(src, dst):
+    """Rename a file from ``src`` to ``dst``.
+
+    We use a custom version rather than the standard library because we
+    have two requirements:
+
+    *   Moves must be atomic.  Otherwise Loris may serve a partial image from
+        a cache, which causes an error.  ``shutil.move()`` is not atomic.
+    *   Moves must work across filesystems.  Often temp directories and the
+        cache directories live on different filesystems.  ``os.rename()`` can
+        throw errors if run across filesystems.
+
+    So we try ``os.rename()``, but if we detect a cross-filesystem copy, we
+    switch to ``shutil.move()`` with some wrappers to make it atomic.
+    """
+    try:
+        os.rename(src, dst)
+    except OSError as err:
```

Comment:
Perhaps obvious, but there is a performance/complexity penalty of catching the exception in the cross-filesystem case. I don't think it's a problem because the movement of a image file between filesystem will dwarf any local optimization made here.

Reply:
Mmm. Assuming we don’t watch to check this every time, we could have an in-memory cache that maps `(src, dst)` to a boolean, recording whether we got an error. If we got an error previously, we skip the `rename(2) `call, and go straight to using `shutil.move()`. Does that sound reasonable?

Reply:
do we have any numbers on performance issues with a try/except? Is it worth the complexity of creating a cache? If the try/except is an issue, another possibility would be to add a setting that skips the check and goes right to the shutil.move (might be simpler than maintaining a cache, but of course then the user would have to set the setting to True).

Reply:
I don’t like the idea of making it a setting, it’s not something the user should worry about. If we go down that route, we can detect it and cache it – this will only be used on a handful of directories.

I’m busy this week with PyCon UK, but assuming this isn’t too urgent, I can do a bit of perf testing sometime next week.

Reply:
I just did some very basic performance testing – rename 10k files from one directory to another, across a filesystem boundary. Two cases:

1. The code as presented here, without a cache
2. Skipping straight into the `shutil.move`, `os.rename` dance – simulating a warm cache that knows not to try `os.rename` in one go.

Time different is negligible, so I think I’ll leave it without a cache.